readme.rst
IwHTTP for Marmalade
==================

This module can be used to work with standard http and https 
protocols



mdev/changelog.txt
iwhttp (1.0.0-1)

  * Add changlog
  * Add package.yml
  * Add README.rst

 -- Ivan Beliy <ivan.beliy@marmalademail.com>  Tue, 19 Jun 2012 11:05:12 +0000


test/echo_http.py
#!/usr/bin/env python
#
# Copyright (C) 2001-2012 Ideaworks3D Ltd.
# All Rights Reserved.
#
# This document is protected by copyright, and contains information
# proprietary to Ideaworks3D.
# This file consists of source code released by Ideaworks3D under
# the terms of the accompanying End User License Agreement (EULA).
# Please do not use this program/source code before you have read the
# EULA and have agreed to be bound by its terms.
#
import sys

out = ""
while 1:
    line = sys.stdin.readline()
    if line.strip() == "":
        break
    out += line

print "HTTP/1.1 200 OK\r\n"
print "Content-Type: text-plain\r\n"
print "Content-Length: %s\r\n\r\n" % len(out)
print out
print "\r\n"


test/test_Get.cpp
/*
 * Copyright (C) 2001-2012 Ideaworks3D Ltd.
 * All Rights Reserved.
 *
 * This document is protected by copyright, and contains information
 * proprietary to Ideaworks Labs.
 * This file consists of source code released by Ideaworks Labs under
 * the terms of the accompanying End User License Agreement (EULA).
 * Please do not use this program/source code before you have read the
 * EULA and have agreed to be bound by its terms.
 */
#include "IwTest.h"
#include "IwRandom.h"
#include "s3eDevice.h"
#include "test_iwhttp.h"

#define TEST_GROUP IwHTTP

#define EXPECTED "This is an example.\n"

int32 GotData(void*, void*);

std::map<CIwHTTP*, CHttpTest*> CHttpTest::s_Instances;

IW_TEST(test_iwhttp_GetSimple)
{
    CHttpTest test;
    test.m_TheStack.Get("http://www.google.com", CHttpTest::GotHeaders, NULL);
    return test.run();
}

IW_TEST(test_iwhttp_GetMulti)
{
    CHttpTest test;
    test.m_TheStack.Get("http://www.yahoo.com", CHttpTest::GotHeaders, NULL);

    // Also check that it's impossible to do 2 requests at once.
    if (test.m_TheStack.Get("http://www.ask.com", CHttpTest::GotHeaders, NULL) !=
        S3E_RESULT_ERROR)
    {
        IwTestError("Second simultaneous request didn't fail");
        return false;
    }

    return test.run();
}

IW_TEST(test_iwhttp_GetCheckResult)
{
    CHttpTest test;
    test.m_TheStack.Get("http://test.ideaworkslabs.com/example.txt", CHttpTest::GotHeaders, NULL);

    if (!test.run())
        return false;

    std::string ctype;
    test.m_TheStack.GetHeader("Content-Type", ctype);
    if (strcmp("text/plain", ctype.c_str()))
    {
        IwTestError("Wrong charset");
        return false;
    }

    if (test.m_TheStack.GetResponseCode() != 200)
    {
        IwTestError("Wrong result code");
        return false;
    }

    IwTrace(CANARY, ("%s", test.m_TheResponse));
    if (test.m_TheResponse && strncmp(EXPECTED, test.m_TheResponse, strlen(EXPECTED)))
    {
        IwTestError("Wrong data");
        return false;
    }

    return true;
}

IW_TEST(test_iwhttp_GetLargeResponse)
{
    // This gets a file that's a bit over a meg long,
    // thus testing non-trivial response lengths. File is also binary
    // so charset should be iso-8859-1

    CHttpTest test;
    test.m_TheStack.Get("http://test.ideaworkslabs.com/largefile.txt", CHttpTest::GotHeaders, NULL);

    if (!test.run())
        return false;

    std::string ctype;
    test.m_TheStack.GetHeader("Content-Type", ctype);
    if (strcmp("text/plain", ctype.c_str()))
    {
        IwTestError("Wrong charset");
        return false;
    }

    if (test.m_TheStack.GetResponseCode() != 200)
    {
        IwTestError("Wrong result code");
        return false;
    }

    // Check the known length of the file
    if (test.m_TheStack.ContentReceived() != 1386439)
    {
        IwTestError("Wrong size: %d");
        return false;
    }
    return true;
}

/**
 * Repeat test 2 but this time use an encrypted connection.
 */
IW_TEST(test_iwhttp_GetSSL)
{

    CHttpTest test;
    test.m_TheStack.Get("https://test.ideaworkslabs.com/example.txt", CHttpTest::GotHeaders, NULL);

    if (!test.run())
        return false;

    std::string ctype;
    test.m_TheStack.GetHeader("Content-Type", ctype);
    if (strcmp("text/plain", ctype.c_str()))
    {
        IwTestError("Wrong charset: %s != text/plain", ctype.c_str());
        return false;
    }

    if (test.m_TheStack.GetResponseCode() != 200)
    {
        IwTestError("Wrong result code");
        return false;
    }

    IwTrace(HTTP, ("%s", test.m_TheResponse));
    if (test.m_TheResponse && strncmp(EXPECTED, test.m_TheResponse, strlen(EXPECTED)))
    {
        IwTestError("Wrong data");
        return false;
    }

    return true;
}

/**
 * Test that IwHTTP objects can repeatedly be created and destroyed without
 * leaking sockets
 */
IW_TEST(test_iwhttp_GetMany)
{
    const unsigned int _TIMEOUT = 10000;
    for (int i = 0; i < 16; i++)
    {
        CHttpTest test;
        test.m_TheStack.Get("http://test.ideaworkslabs.com/example.txt", CHttpTest::GotHeaders, NULL);
        test.m_Timeout = _TIMEOUT;
        if (!test.run())
            return false;

        if (test.m_TheStack.GetStatus() != S3E_RESULT_SUCCESS || strncmp(EXPECTED, test.m_TheResponse, strlen(EXPECTED)) != 0)
        {
            IwTestError("Failed to fetch data");
            return false;
        }
    }
    return true;
}

/**
 * Test that IwHTTP objects can repeatedly be created and destroyed without
 * leaking sockets except this time for HTTPS://
 */
IW_TEST(test_iwhttp_GetManySSL)
{
    const unsigned int _TIMEOUT = 10000;
    for (int i = 0; i < 16; i++)
    {
        CHttpTest test;
        test.m_TheStack.Get("https://test.ideaworkslabs.com/example.txt", CHttpTest::GotHeaders, NULL);
        test.m_Timeout = _TIMEOUT;
        if (!test.run())
            return false;

        if (test.m_TheStack.GetStatus() != S3E_RESULT_SUCCESS || strncmp(EXPECTED, test.m_TheResponse, strlen(EXPECTED)) != 0)
        {
            IwTestError("Failed to fetch data");
            return false;
        }
    }
    return true;
}

/**
 * Test chunked-encoding with standard chunks..
 */
IW_TEST(test_iwhttp_GetChunked)
{
    int buf_sizes[] = { 10, 1000 };

    // Repeat this test 3 times with different read buffer sizes
    for (uint i = 0; i < sizeof(buf_sizes)/sizeof(int); i++)
    {
        CHttpTest test;

        // This URL will return Transfer-Encoded: chunked
        test.m_TheStack.Get("http://test.ideaworkslabs.com/index.php", CHttpTest::GotHeaders, (void*)true);
        if (!test.getHeaders())
            return false;

        std::string ctype;
        test.m_TheStack.GetHeader("Content-Type", ctype);
        if (strcmp("text/html", ctype.c_str()))
        {
            IwTestError("Wrong charset");
            return false;
        }

        if (test.m_TheStack.GetResponseCode() != 200)
        {
            IwTestError("Wrong result code");
            return false;
        }

        int32 bufSize = buf_sizes[i];

        // Read the chunked body..
        std::string reply;
        char* buf = new char[bufSize + 1];
        while (!test.m_TheStack.ContentFinished() && !test.checkTimeout())
        {
            memset(buf, 0, bufSize + 1);
            int bytes_read = test.m_TheStack.ReadContent(buf, bufSize, NULL, NULL);
            if (bytes_read > bufSize)
            {
                IwTestError("Read too much");
                return false;
            }
            reply += buf;
            s3eDeviceYield(0);
        }
        delete buf;

        // Although the received data is variable, the last bytes are always the PHP license and closing HTML
        const char *expected = "<p>If you did not receive a copy of the PHP license, or have any questions about PHP licensing, please contact license@php.net.\n" \
        "</p>\n</td></tr>\n</table><br />\n</div></body></html>1"; // Server always sends that last '1' ??

        if (reply.find(expected) != reply.size() - strlen(expected))
        {
            IwTestError(
                "Did not read all chunked data. BufSize: %d Timeout: %d Recvd: %d ReplySize: %d",
                bufSize, test.checkTimeout(), test.m_TheStack.ContentReceived(), reply.size()
            );
            return false;
        }

        if (test.m_Status == S3E_RESULT_ERROR)
        {
            IwTestError("Fetch failed");
            return false;
        }
    }

    return true;
}

/**
 * Test chunked-encoding with tiny chunks..
 */
IW_TEST(test_iwhttp_GetChunkTiny)
{
    int buf_sizes[] = { 1, 10, 1000 };

    int i;
    // Repeat this test 3 times with different read buffer sizes
    for (i = 0; i < 3; i++)
    {
        CHttpTest test;

        // This URL will return Transfer-Encoded: chunked
        test.m_TheStack.Get("http://test.ideaworkslabs.com/smallchunks.php", CHttpTest::GotHeaders, (void*)true);

        if (!test.getHeaders())
            return false;

        std::string ctype;
        test.m_TheStack.GetHeader("Content-Type", ctype);
        if (strcmp("text/html", ctype.c_str()))
        {
            IwTestError("Wrong charset");
            return false;
        }

        if (test.m_TheStack.GetResponseCode() != 200)
        {
            IwTestError("Wrong result code");
            return false;
        }

        IwRandSeed(time(NULL) & 0xFFFFFFFF);
        int32 bufSize = buf_sizes[i];

        // Read the chunked body..
        std::string reply;
        char *buf = new char[bufSize + 1];
        while (!test.m_TheStack.ContentFinished() && !test.checkTimeout())
        {
            memset(buf, 0, bufSize + 1);
            int bytes_read = test.m_TheStack.ReadContent(buf, bufSize, NULL, NULL);
            if (bytes_read > bufSize)
            {
                IwTestError("Read too much");
                return false;
            }

            reply += buf;

            s3eDeviceYield(10);
        }
        delete buf;

        // Although the received data is variable, the last bytes are always the PHP license and closing HTML
        const char *expected = "789";

        if (reply.find(expected) != reply.size() - strlen(expected))
        {
            IwTestError("Did not read all chunked data. BufSize: %d Timeout: %d Recvd: %d", bufSize,
                        test.checkTimeout(),
                        test.m_TheStack.ContentReceived());
            return false;
        }
    }

    return true;
}

/**
 * Test chunked-encoding with standard chunks with HTTPS
 */
IW_TEST(test_iwhttp_GetChunkSSL)
{
    int buf_sizes[] = { 10, 1000 };

    // Repeat this test 3 times with different read buffer sizes
    for (uint i = 0; i < sizeof(buf_sizes)/sizeof(int); i++)
    {
        CHttpTest test;

        // This URL will return Transfer-Encoded: chunked
        test.m_TheStack.Get("https://test.ideaworkslabs.com/index.php", CHttpTest::GotHeaders, (void*)true);

        if (!test.getHeaders())
            return false;

        std::string ctype;
        test.m_TheStack.GetHeader("Content-Type", ctype);
        if (strcmp("text/html", ctype.c_str()))
        {
            IwTestError("Wrong charset");
            return false;
        }

        if (test.m_TheStack.GetResponseCode() != 200)
        {
            IwTestError("Wrong result code");
            return false;
        }

        int32 bufSize = buf_sizes[i];

        // Read the chunked body..
        std::string reply;
        char *buf = new char[bufSize + 1];
        while (!test.m_TheStack.ContentFinished() && !test.checkTimeout())
        {
            memset(buf, 0, bufSize + 1);
            test.m_TheStack.ReadContent(buf, bufSize, NULL, NULL);
            reply += buf;

            s3eDeviceYield(0);
        }
        delete buf;

        // Although the received data is variable, the last bytes are always the PHP license and closing HTML
        const char *expected = "<p>If you did not receive a copy of the PHP license, or have any questions about PHP licensing, please contact license@php.net.\n" \
        "</p>\n</td></tr>\n</table><br />\n</div></body></html>1"; // Server always sends that last '1' ??

        if (reply.find(expected) != reply.size() - strlen(expected))
        {
            IwTestError(
                "Did not read all chunked data. BufSize: %d Timeout: %d Recvd: %d ReplySize: %d",
                bufSize, test.checkTimeout(), test.m_TheStack.ContentReceived(), reply.size()
            );
            return false;
        }
    }

    return true;
}

/**
 * Test chunked-encoding with HTTPS with tiny chunks..
 */
IW_TEST(test_iwhttp_TinyChunkSSL)
{
    int buf_sizes[] = { 1, 10, 1000 };

    // Repeat this test 3 times with different read buffer sizes
    for (uint i = 0; i < sizeof(buf_sizes)/sizeof(int); i++)
    {
        CHttpTest test;

        // This URL will return Transfer-Encoded: chunked
        test.m_TheStack.Get("https://test.ideaworkslabs.com/smallchunks.php", CHttpTest::GotHeaders, (void*)true);

        if (!test.getHeaders())
            return false;

        std::string ctype;
        test.m_TheStack.GetHeader("Content-Type", ctype);
        if (strcmp("text/html", ctype.c_str()))
        {
            IwTestError("Wrong charset");
            return false;
        }

        if (test.m_TheStack.GetResponseCode() != 200)
        {
            IwTestError("Wrong result code");
            return false;
        }

        IwRandSeed(time(NULL) & 0xFFFFFFFF);
        int32 bufSize = buf_sizes[i];

        // Read the chunked body..
        std::string reply;
        char *buf = new char[bufSize + 1];
        while (!test.m_TheStack.ContentFinished() && !test.checkTimeout())
        {
            memset(buf, 0, bufSize + 1);
            test.m_TheStack.ReadContent(buf, bufSize, NULL, NULL);
            reply += buf;

            s3eDeviceYield(10);
        }
        delete buf;

        const char *expected = "0123456789";
        if (reply.find(expected) != reply.size() - strlen(expected))
        {
            IwTestError("Did not read all chunked data. BufSize: %d Timeout: %d Recvd: %d", bufSize, test.checkTimeout(), test.m_TheStack.ContentReceived());
            return false;
        }
    }

    return true;
}

/**
 * A test for blocking reads, smallchunksdelay.php deliberately
 * stalls when sending, non-SSL version
 */
IW_TEST(test_iwhttp_GetBlocking)
{
    CHttpTest test;
    //int i = 0;

    // This URL will return Transfer-Encoded: chunked
    test.m_TheStack.Get("http://test.ideaworkslabs.com/smallchunksdelay.php", CHttpTest::GotHeaders, (void*)true);

    if (!test.getHeaders())
        return false;

    std::string ctype;
    test.m_TheStack.GetHeader("Content-Type", ctype);
    if (strcmp("text/html", ctype.c_str()))
    {
        IwTestError("Wrong charset");
        return false;
    }

    if (test.m_TheStack.GetResponseCode() != 200)
    {
        IwTestError("Wrong result code");
        return false;
    }

    IwRandSeed(time(NULL) & 0xFFFFFFFF);
    int bufSize = 10;

    // Read the chunked body..
    std::string reply;
    char *buf = new char[bufSize + 1];
    while (!test.m_TheStack.ContentFinished() && !test.checkTimeout())
    {
        memset(buf, 0, bufSize + 1);
        test.m_TheStack.ReadContent(buf, bufSize, NULL, NULL);
        reply += buf;

        s3eDeviceYield(10);
    }
    delete buf;

    const char *expected = "0123456789";
    if (reply.find(expected) != reply.size() - strlen(expected))
    {
        IwTestError(
            "Did not read all chunked data. BufSize: %d Timeout: %d Recvd: %d", bufSize, test.checkTimeout(), test.m_TheStack.ContentReceived()
        );
        return false;
    }

    return true;
}

/**
 * A test for blocking reads, smallchunksdelay.php deliberately
 * stalls when sending, SSL version
 */
IW_TEST(test_iwhttp_GetBlockingSSL)
{
    CHttpTest test;

    // This URL will return Transfer-Encoded: chunked
    test.m_TheStack.Get("https://test.ideaworkslabs.com/smallchunksdelay.php", CHttpTest::GotHeaders, (void*)true);
    if (!test.getHeaders())
        return false;

    std::string ctype;
    test.m_TheStack.GetHeader("Content-Type", ctype);
    if (strcmp("text/html", ctype.c_str()))
    {
        IwTestError("Wrong charset");
        return false;
    }

    if (test.m_TheStack.GetResponseCode() != 200)
    {
        IwTestError("Wrong result code");
        return false;
    }

    IwRandSeed(time(NULL) & 0xFFFFFFFF);
    int32 bufSize = 10;

    // Read the chunked body..
    std::string reply;
    char *buf = new char[bufSize + 1];
    while (!test.m_TheStack.ContentFinished() && !test.checkTimeout())
    {
        memset(buf, 0, bufSize + 1);
        test.m_TheStack.ReadContent(buf, bufSize, NULL, NULL);
        reply += buf;

        s3eDeviceYield(10);
    }
    delete buf;

    const char *expected = "0123456789";
    if (reply.find(expected) != reply.size() - strlen(expected))
    {
        IwTestError(
            "Did not read all chunked data. BufSize: %d Timeout: %d Recvd: %d", bufSize, test.checkTimeout(), test.m_TheStack.ContentReceived()
        );
        return false;
    }

    if (test.m_Status == S3E_RESULT_ERROR)
    {
        IwTestError("Fetch failed");
        return false;
    }

    return true;
}

static int32 GotHeaders_Get13(void *sysData, void *usrData)
{
    CIwHTTP *http = (CIwHTTP*)sysData;
    bool *ret = (bool*)usrData;

    IwTestTrace("GotHeaders_Get13: %d %d", http->ContentExpected(), http->ContentFinished());
    if (http->ContentExpected() > 100 || http->ContentFinished() == true)
        *ret = false;
    else
        *ret = true;

    CHttpTest::GetInstance(http)->HandleHeaders(true);
    return 0;
}

/**
 * A test for a common usage of GotHeaders callback under chunked-encoding
 * which causes lots of confusion amongst users..
 */
IW_TEST(test_iwhttp_ChunkedHeaders)
{
    CHttpTest test;
    bool ret = false;

    // This URL will return Transfer-Encoded: chunked
    test.m_TheStack.Get("http://test.ideaworkslabs.com/smallchunksdelay.php", GotHeaders_Get13, (void*)&ret);

    if (!test.getHeaders())
        return false;

    std::string ctype;
    test.m_TheStack.GetHeader("Content-Type", ctype);
    if (strcmp("text/html", ctype.c_str()))
    {
        IwTestError("Wrong charset");
        return false;
    }

    if (test.m_TheStack.GetResponseCode() != 200)
    {
        IwTestError("Wrong result code");
        return false;
    }

    int32 bufSize = 10;

    // Read the chunked body..
    std::string reply;
    char *buf = new char[bufSize + 1];
    while (!test.m_TheStack.ContentFinished() && !test.checkTimeout())
    {
        memset(buf, 0, bufSize + 1);
        test.m_TheStack.ReadContent(buf, bufSize, NULL, NULL);
        reply += buf;

        s3eDeviceYield(10);

        if (test.m_TheStack.ContentReceived() > test.m_TheStack.ContentExpected())
            break;

        if (test.m_TheStack.ContentReceived() != reply.size())
            break;
    }
    delete buf;

    if (test.m_TheStack.ContentReceived() != reply.size())
    {
        IwTestError(
            "Received totals don't match: %d %d", test.m_TheStack.ContentReceived(), reply.size()
        );
        return false;
    }

    if (test.m_TheStack.ContentReceived() != test.m_TheStack.ContentExpected())
    {
        IwTestError(
            "Content totals don't match: %d %d", test.m_TheStack.ContentReceived(), test.m_TheStack.ContentExpected()
        );
        return false;
    }

    const char *expected = "0123456789";
    if (reply.find(expected) != reply.size() - strlen(expected))
    {
        IwTestError(
            "Did not read all chunked data. BufSize: %d Timeout: %d Recvd: %d", bufSize, test.checkTimeout(), test.m_TheStack.ContentReceived()
        );
        return false;
    }
    return ret;
}

/**
 * A test for a common usage of GotHeaders callback under chunked-encoding
 * which causes lots of confusion amongst users.. SSL version
 */
IW_TEST(test_iwhttp_ChunkedHeadersSSL)
{
    CHttpTest test;
    bool ret = false;

    // This URL will return Transfer-Encoded: chunked
    test.m_TheStack.Get("https://test.ideaworkslabs.com/smallchunksdelay.php", GotHeaders_Get13, (void*)&ret);

    if (!test.getHeaders())
        return false;

    std::string ctype;
    test.m_TheStack.GetHeader("Content-Type", ctype);
    if (strcmp("text/html", ctype.c_str()))
    {
        IwTestError("Wrong charset");
        return false;
    }

    if (test.m_TheStack.GetResponseCode() != 200)
    {
        IwTestError("Wrong result code");
        return false;
    }

    int bufSize = 100;

    // Read the chunked body..
    std::string reply;
    char *buf = new char[bufSize + 1];
    while (!test.m_TheStack.ContentFinished() && !test.checkTimeout())
    {
        memset(buf, 0, bufSize + 1);
        test.m_TheStack.ReadContent(buf, bufSize, NULL, NULL);
        reply += buf;

        s3eDeviceYield(10);

        if (test.m_TheStack.ContentReceived() > test.m_TheStack.ContentExpected())
            break;

        if (test.m_TheStack.ContentReceived() != reply.size())
            break;
    }
    delete buf;

    if (test.m_TheStack.ContentReceived() != reply.size())
    {
        IwTestError(
            "Received totals don't match: %d %d", test.m_TheStack.ContentReceived(), reply.size()
        );
        return false;
    }

    if (test.m_TheStack.ContentReceived() != test.m_TheStack.ContentExpected())
    {
        IwTestError(
            "Content totals don't match: %d %d", test.m_TheStack.ContentReceived(), test.m_TheStack.ContentExpected()
        );
        return false;
    }

    const char *expected = "0123456789";
    if (reply.find(expected) != reply.size() - strlen(expected))
    {
        IwTestError(
            "Did not read all chunked data. BufSize: %d Timeout: %d Recvd: %d", bufSize, test.checkTimeout(), test.m_TheStack.ContentReceived()
        );
        return false;
    }

    if (test.m_Status == S3E_RESULT_ERROR)
    {
        IwTestError("Fetch failed");
        return false;
    }
    return ret;
}


test/test_Uri.cpp
/*
 * Copyright (C) 2001-2012 Ideaworks3D Ltd.
 * All Rights Reserved.
 *
 * This document is protected by copyright, and contains information
 * proprietary to Ideaworks Labs.
 * This file consists of source code released by Ideaworks Labs under
 * the terms of the accompanying End User License Agreement (EULA).
 * Please do not use this program/source code before you have read the
 * EULA and have agreed to be bound by its terms.
 */
#include "IwURI.h"
#include "IwTest.h"

#define TEST_GROUP IwHTTP

#define URI1 "http://www.ideaworks3d.com"
#define URI2 "ftp://www.ideaworks3d.com/somestuff?query"
#define URI3 "thing://www.ideaworks3d.com:81"
#define URI4 "https://www.ideaworks3d.com?query"
#define URI5 "file:///stuff"
#define URI6 "http://www.ideaworks3d.com:81/path?query"
#define URI7 "mailto:abuse@ideaworks3d.com"

IW_TEST(test_iwhttp_HostPresent1)
{
    CIwURI t(URI1);
    if (strcmp(t.GetHost(), "www.ideaworks3d.com"))
        return false;
    return true;
}

IW_TEST(test_iwhttp_HostPresent2)
{
    CIwURI s(URI2);
    CIwURI t("http://www.google.com");
    t = s;
    if (strcmp(t.GetHost(), "www.ideaworks3d.com"))
        return false;
    return true;
}
IW_TEST(test_iwhttp_HostPresent3)
{
    CIwURI s(URI3);
    CIwURI t(s);
    if (strcmp(t.GetHost(), "www.ideaworks3d.com"))
        return false;
    return true;
}

IW_TEST(test_iwhttp_HostPresent4)
{
    CIwURI t;
    t = URI4;
    if (strcmp(t.GetHost(), "www.ideaworks3d.com"))
        return false;
    return true;
}

IW_TEST(test_iwhttp_HostPresent5)
{
    CIwURI t(URI5);
    if (t.GetHost())
        return false;
    return true;
}

IW_TEST(test_iwhttp_HostPresent6)
{
    CIwURI t(URI6);
    if (strcmp(t.GetHost(), "www.ideaworks3d.com"))
        return false;
    return true;
}

IW_TEST(test_iwhttp_HostPresent7)
{
    CIwURI t(URI5);
    if (t.GetHost())
        return false;
    return true;
}

IW_TEST(test_iwhttp_Scheme1)
{
    CIwURI t(URI1);
    if (t.GetProtocol() != CIwURI::HTTP)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Scheme2)
{
    CIwURI t(URI2);
    if (t.GetProtocol() != CIwURI::FTP)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Scheme3)
{
    CIwURI t(URI3);
    if (t.GetProtocol() != CIwURI::UNKNOWN)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Scheme4)
{
    CIwURI t(URI4);
    if (t.GetProtocol() != CIwURI::HTTPS)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Scheme5)
{
    CIwURI t(URI5);
    if (t.GetProtocol() != CIwURI::FILE)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Scheme6)
{
    CIwURI t(URI6);
    if (t.GetProtocol() != CIwURI::HTTP)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Scheme7)
{
    CIwURI t(URI7);
    if (t.GetProtocol() != CIwURI::UNKNOWN)
        return false;
    return true;
}


IW_TEST(test_iwhttp_Port1)
{
    CIwURI t(URI1);
    if (t.GetPort() != 80)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Port2)
{
    CIwURI t(URI2);
    if (t.GetPort() != 21)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Port3)
{
    CIwURI t(URI3);
    if (t.GetPort() != 81)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Port4)
{
    CIwURI t(URI4);
    if (t.GetPort() != 443)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Port5)
{
    CIwURI t(URI5);
    if (t.GetPort() != 0)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Port6)
{
    CIwURI t(URI6);
    if (t.GetPort() != 81)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Port7)
{
    CIwURI t(URI7);
    if (t.GetPort() != 0)
        return false;
    return true;
}

IW_TEST(test_iwhttp_Tail1)
{
    CIwURI t(URI1);
    if (t.GetTail())
        return false;
    return true;
}

IW_TEST(test_iwhttp_Tail2)
{
    CIwURI t(URI2);
    if (strcmp(t.GetTail(), "somestuff?query"))
        return false;
    return true;
}

IW_TEST(test_iwhttp_Tail3)
{
    CIwURI t(URI3);
    if (t.GetTail())
        return false;
    return true;
}

IW_TEST(test_iwhttp_Tail4)
{
    CIwURI t(URI4);
    if (strcmp(t.GetTail(), "?query"))
        return false;
    return true;
}

IW_TEST(test_iwhttp_Tail5)
{
    CIwURI t(URI5);
    if (strcmp(t.GetTail(), "stuff"))
        return false;
    return true;
}

IW_TEST(test_iwhttp_Tail6)
{
    CIwURI t(URI6);
    if (strcmp(t.GetTail(), "path?query"))
        return false;
    return true;
}

IW_TEST(test_iwhttp_Tail7)
{
    CIwURI t(URI7);
    if (t.GetTail())
        return false;
    return true;
}

IW_TEST(test_iwhttp_All1)
{
    CIwURI t(URI1);
    if (strcmp(t.GetAll(), URI1))
        return false;
    return true;
}

IW_TEST(test_iwhttp_All2)
{
    CIwURI t(URI2);
    if (strcmp(t.GetAll(), URI2))
        return false;
    return true;
}

IW_TEST(test_iwhttp_All3)
{
    CIwURI t(URI3);
    if (strcmp(t.GetAll(), URI3))
        return false;
    return true;
}

IW_TEST(test_iwhttp_All4)
{
    CIwURI t(URI4);
    if (strcmp(t.GetAll(), URI4))
        return false;
    return true;
}

IW_TEST(test_iwhttp_All5)
{
    CIwURI t(URI5);
    if (strcmp(t.GetAll(), URI5))
        return false;
    return true;
}

IW_TEST(test_iwhttp_All6)
{
    CIwURI t(URI6);
    if (strcmp(t.GetAll(), URI6))
        return false;
    return true;
}

IW_TEST(test_iwhttp_All7)
{
    CIwURI t(URI7);
    if (strcmp(t.GetAll(), URI7))
        return false;
    return true;
}


h/IwUriEscape.h
/*
 * Copyright (C) 2001-2012 Ideaworks3D Ltd.
 * All Rights Reserved.
 *
 * This document is protected by copyright, and contains information
 * proprietary to Ideaworks Labs.
 * This file consists of source code released by Ideaworks Labs under
 * the terms of the accompanying End User License Agreement (EULA).
 * Please do not use this program/source code before you have read the
 * EULA and have agreed to be bound by its terms.
 */

#ifndef IW_URI_ESCAPE_H
#define IW_URI_ESCAPE_H

#include <string>

/**
 * @addtogroup iwhttpgroup
 * @{
 */

/// A class for URI-escaping functions.
class CIwUriEscape
{
public:
    /// URI-escapes a string, for use in the middle of a query.
    /// in other words, the text is turned into something%20like%20this.
    /// @param in The string to escape
    /// @return The escaped string
    static std::string Escape(std::string in);
};

/** @} */
#endif


src/IwUriEscape.cpp
/*
 * Copyright (C) 2001-2012 Ideaworks3D Ltd.
 * All Rights Reserved.
 *
 * This document is protected by copyright, and contains information
 * proprietary to Ideaworks Labs.
 * This file consists of source code released by Ideaworks Labs under
 * the terms of the accompanying End User License Agreement (EULA).
 * Please do not use this program/source code before you have read the
 * EULA and have agreed to be bound by its terms.
 */

#include "IwUriEscape.h"

std::string CIwUriEscape::Escape(std::string in)
{
    std::string ret;
    for (std::string::iterator it = in.begin(); it != in.end(); it++)
    {
        char c = *it;
        if (isalnum(c))
        {
            ret += c;
        }
        else
        {
            char buf[5];
            snprintf(buf, 5, "%%%.2x", c);
            ret += buf;
        }
    }
    return ret;
}


src/IwURI.cpp
/*
 * Copyright (C) 2001-2012 Ideaworks3D Ltd.
 * All Rights Reserved.
 *
 * This document is protected by copyright, and contains information
 * proprietary to Ideaworks Labs.
 * This file consists of source code released by Ideaworks Labs under
 * the terms of the accompanying End User License Agreement (EULA).
 * Please do not use this program/source code before you have read the
 * EULA and have agreed to be bound by its terms.
 */

#include "IwURI.h"

#include <ctype.h>
#include <string.h>
#include <stdlib.h>

CIwURI::CIwURI() : m_pURI(NULL)
{
}

CIwURI::CIwURI(const char *URI) : m_pURI(NULL)
{
    *this = URI;
}

CIwURI::CIwURI(const CIwURI& URI) : m_pURI(NULL)
{
    *this = URI;
}

CIwURI::~CIwURI()
{
    if (m_pURI)
        delete m_pURI;
}

CIwURI& CIwURI::operator=(const char *URI)
{
    if (m_pURI)
        delete m_pURI;

    m_pURI = new char[strlen(URI) + 1];
    strcpy(m_pURI, URI);

    ParseURI();

    return *this;
}

CIwURI& CIwURI::operator=(const CIwURI& URI)
{
    *this = URI.GetAll();
    return *this;
}

void CIwURI::ParseURI()
{
    struct _PROTOCOLS { const char *p;    CIwURI::PROTOCOL e; uint16 port;    };
    static _PROTOCOLS prots[] = {
        { "http://", HTTP, 80 },
        { "https://", HTTPS, 443 },
        { "ftp://", FTP, 21 },
        { "file://", FILE, 0 },
        { NULL, UNKNOWN }
    };


    m_pHost = NULL;
    m_pTail = NULL;
    m_protocol = UNKNOWN;
    m_pRemoved = NULL;
    m_port = 0;

    unsigned int uri_len = strlen(m_pURI);

    struct _PROTOCOLS *prot = &prots[0];
    while (prot->p)
    {
        unsigned int len = strlen(prot->p);

        if (strncmp(prot->p, m_pURI, len) == 0)
        {
            m_pHost = &m_pURI[len];
            m_protocol = prot->e;
            m_port = prot->port;
            break;
        }
        prot++;
    }

    if (!m_pHost)
    {
        // We didn't recognise the protocol but should
        // still try to find the host part.

        char *delim = strstr(m_pURI, "://");
        if (delim && ((delim - m_pURI + 3) < (int)uri_len) )
            m_pHost = &delim[3];
    }

    char *pChar = m_pHost;
    while (pChar && *pChar && *pChar != '/' && *pChar != ':' && *pChar != '?') pChar++;
    if (pChar && *pChar)
    {
        if (*pChar == '/' || *pChar == '?')
        {
            m_pRemoved = pChar;
            m_removed = *m_pRemoved;
            *pChar = 0; // Null-terminate host
        }
        else if (*pChar == ':')
        {
            m_pRemoved = pChar;
            m_removed = *m_pRemoved;
            *pChar = 0; // Null-terminate host
            m_port = atoi(++pChar);
            while (isdigit(*pChar)) pChar++;
        }

        if (m_removed == '?')
            m_pTail = pChar;
        else if ((m_pRemoved == pChar && m_removed) || *pChar) // If the character that was at pChar is not NULL
            m_pTail = ++pChar; // Then there's a tail
    }
    if (m_pHost && !*m_pHost) // If the host is empty, claim it's not there
    {
        m_pHost = NULL;
    }
}

void CIwURI::TerminateHost() const
{
    if (m_pRemoved && *m_pRemoved)
    {
        *m_pRemoved = 0;
    }
}

void CIwURI::UnterminateHost() const
{
    if (m_pRemoved && !*m_pRemoved)
    {
        *m_pRemoved = m_removed;
    }
}

const char *CIwURI::GetAll() const
{
    UnterminateHost();
    return m_pURI;
}
